dev/tools/controllerbuilder/pkg/updatetypes/main.go

```diff
@@ -0,0 +1,235 @@
+func findNewField(pbSourcePath, parentMsgFullName, newFieldName string) (protoreflect.MessageDescriptor, protoreflect.FieldDescriptor, error) {
+	fileData, err := os.ReadFile(pbSourcePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %q: %w", pbSourcePath, err)
+	}
+
+	fds := &descriptorpb.FileDescriptorSet{}
+	if err := proto.Unmarshal(fileData, fds); err != nil {
+		return nil, nil, fmt.Errorf("unmarshalling %q: %w", pbSourcePath, err)
+	}
+
+	files, err := protodesc.NewFiles(fds)
+	if err != nil {
+		return nil, nil, fmt.Errorf("building file description: %w", err)
+	}
+
+	// Find the parent message
+	messageDesc, err := files.FindDescriptorByName(protoreflect.FullName(parentMsgFullName))
+	if err != nil {
+		return nil, nil, err
+	}
+	msgDesc, ok := messageDesc.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, nil, fmt.Errorf("unexpected descriptor type: %T", msgDesc)
+	}
+
+	// Find the new field in parent message
+	fieldDesc := msgDesc.Fields().ByName(protoreflect.Name(newFieldName))
+	if fieldDesc == nil {
+		return nil, nil, fmt.Errorf("field not found in message")
+	}
+
+	return msgDesc, fieldDesc, nil
+}
+
+// findDependentMsgs finds all dependent proto messages for the given field, ignoring specified fields
+func findDependentMsgs(field protoreflect.FieldDescriptor, ignoredProtoFields []string) (map[string]protoreflect.MessageDescriptor, error) {
+	deps := make(map[string]protoreflect.MessageDescriptor)
+	analyzeDependencies(field, deps, ignoredProtoFields)
+
+	// fmt.Println("\nDependencies:")
+	// for _, dep := range deps {
+	// 	fmt.Printf("%s\n", dep.FullName())
+	// }
+	return deps, nil
+}
+
+// analyzeDependencies recursively explores the dependent proto messages of the given field.
+func analyzeDependencies(field protoreflect.FieldDescriptor, deps map[string]protoreflect.MessageDescriptor, ignoreFields []string) {
+	if shouldIgnore(string(field.FullName()), ignoreFields) {
+		return
+	}
+
+	if field.IsMap() {
+		/* an example Map field
+		message ParentMessage {
+			map<KeyMessage, ValueMessage> some_map_field = 1;
+		} */
+		mapEntry := field.Message()
+		if keyField := mapEntry.Fields().ByName("key"); keyField != nil {
+			analyzeDependencies(keyField, deps, ignoreFields)
+		}
+		if valueField := mapEntry.Fields().ByName("value"); valueField != nil {
+			analyzeDependencies(valueField, deps, ignoreFields)
+		}
+	} else {
+		switch field.Kind() {
+		case protoreflect.MessageKind:
+			msg := field.Message()
+			fqn := string(msg.FullName())
+			if _, ok := deps[fqn]; !ok {
+				deps[fqn] = msg
+				for i := 0; i < msg.Fields().Len(); i++ {
+					field := msg.Fields().Get(i)
+					analyzeDependencies(field, deps, ignoreFields)
+				}
+			}
+		case protoreflect.EnumKind:
+			// deps[string(field.Enum().FullName())] = true  // Skip enum because enum is mapped to Go string in code generation
+		}
+	}
+}
+
+func shouldIgnore(fieldName string, ignoreFields []string) bool {
```

Comment:
Tip: might be easier to pass a `StringSet` or a `Set[string]`, we already use the one from apimachinery in a few places.

Reply:
Thanks! Address in #2373